plugins/rplanners/parabolicsmoother2.cpp

```diff
@@ -3671,6 +3710,88 @@ class ParabolicSmoother2 : public PlannerBase, public RampOptimizer::Feasibility
+        /// \param[in] pUsedBody : used kinbody.
+        /// \param[in] vUsedDOFIndices : used dof indices for config.
+        void Init(const OpenRAVE::KinBodyConstPtr& pUsedBody, const std::vector<int>& vUsedDOFIndices)
+        {
+            _pUsedBody = pUsedBody;
+            _vUsedDOFIndices = vUsedDOFIndices;
+            InitializeCachedVectors(_vFullDOFPositions, _vFullDOFVelocities, _vFullDOFAccelerationLimits, _vFullDOFJerkLimits, _pUsedBody->GetDOF());
+        }
+
+        /// \brief update limits by dynamic limits. for now, update only acceleration limits.
+        ///        acceleration limits should at least satisfy the dynamic acceleration limits at t0 and t1, which are start/end boundary conditions of trajectory segment.
+        ///        all vectors in arguments have same size and order config, like _parameters->_vConfigVelocityLimit.
+        /// \param[out/in] vAccelLimits : resultant acceleration limits. expected to have the initial acceleration limits in it and this function updates them.
+        /// \param[in] vVelocityLimits : just in case, clamp velocity by velocity limits, since dynamic limit might be ill-condition.
+        /// \param[in] x0Vect, v0Vect, x1Vect, v1Vect : boundary conditions of positions and velocities at t0 and t1. Same size and order as _parameters->_vConfigVelocityLimit.
+        /// \param[in] usedBody : used kinbody, which should support GetDOFDynamicAccelerationJerkLimits API.
+        void UpdateLimitsByDynamicLimits(std::vector<dReal>& vAccelLimits,
+                                         const std::vector<dReal>& vVelocityLimits,
+                                         const std::vector<dReal>& x0Vect, const std::vector<dReal>& x1Vect,
+                                         const std::vector<dReal>& v0Vect, const std::vector<dReal>& v1Vect)
+        {
+            const OpenRAVE::KinBody& usedBody = *_pUsedBody;
+            // check and update dynamic acceleration limit at x0
+            _UpdateLimitsByDynamicLimitsAtBoundary(vAccelLimits, vVelocityLimits, x0Vect, v0Vect, usedBody);
+            // check and update dynamic acceleration limit at x1
+            _UpdateLimitsByDynamicLimitsAtBoundary(vAccelLimits, vVelocityLimits, x1Vect, v1Vect, usedBody);
+        }
+
+        /// \brief initialize cached vector.
+        static void InitializeCachedVectors(std::vector<OpenRAVE::dReal>& vFullDOFPositions,
+                                            std::vector<OpenRAVE::dReal>& vFullDOFVelocities,
+                                            std::vector<OpenRAVE::dReal>& vFullDOFAccelerationLimits,
+                                            std::vector<OpenRAVE::dReal>& vFullDOFJerkLimits,
+                                            const int nDOF)
+        {
+            vFullDOFPositions.assign(nDOF, 0.0);
+            vFullDOFVelocities.assign(nDOF, 0.0);
+            vFullDOFAccelerationLimits.resize(nDOF);
+            vFullDOFJerkLimits.resize(nDOF);
+        }
```

Comment:
I would do the following instead.
- There is no guarantee that the value 0 is a valid joint value.
- Because these dof values are part of inputs for getting the correct dynamic limits, we should set them to the correct values.

```suggestion
        /// \brief initialize cached vectors.
        static void InitializeCachedVectors(const OpenRAVE::KinBodyConstPtr& pBody,
                                            std::vector<OpenRAVE::dReal>& vFullDOFPositions,
                                            std::vector<OpenRAVE::dReal>& vFullDOFVelocities,
                                            std::vector<OpenRAVE::dReal>& vFullDOFAccelerationLimits,
                                            std::vector<OpenRAVE::dReal>& vFullDOFJerkLimits)
        {
            pBody->GetDOFValues(vFullDOFPositions);
            const int nFullDOF = pBody->GetDOF();
            vFullDOFVelocities.assign(nFullDOF, 0.0);
            vFullDOFAccelerationLimits.resize(nFullDOF);
            vFullDOFJerkLimits.resize(nFullDOF);
        }
```

Reply:
thanks!

sounds good!